### DISCUSSION TITLE
flatbread-editor

### DISCUSSION THREAD

**Discussion by odama626** [FlatbreadLabs/flatbread#39] [Ideas]
I've been continuing to brainstorm about the possibility of creating an editor that can accompany flatbread.

since graphql allows for introspection the editor can be pretty generic and use the introspection to dynamically generate the ui and use mutations to update the data

```graphql
query EditorQuery {
  __schema {
    types {
      name
      kind
    }
  }
  __type(name: "Post") {
    name
    fields {
      name
      description
      type {
        kind
        name
        ofType {
          name
          kind
        }
      }
    }
  }
}
```

For the editor to work we need to allow updating and creating new records
- core should allow for mutations that can update or create new records
- transformers should be able to handle converting data back into a VFile
- sources should be able to update and create new sources
- resolvers can  export a json object that explain custom types for the editor, or possibly a payload that can be imported as a widget 

we could also allow for a netlify style yaml config to improve the editor view of fields if needed.
- date fields that are expressed as strings can specify a date widget
- fields can have an added description or enforced min / max for validation
- fields can be marked hidden, read only, or provide a resolver that generates a default value?

**Comment by odama626** [comment]
I started thinking more about this and realized we don't currently have a plan for how to handle creating and storing the editor configs, is that something that should be done through graphql through a source?

**Reply by tonyketcham** [reply]
Yeah I think that's the best approach. That would essentially be an extended method of what Netlify CMS does
